File: android/app/src/main/kotlin/com/example/nomatch/OrientationVectorManager.kt
package com.example.nomatch

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.util.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import kotlin.math.*

class OrientationVectorManager(
    private val context: Context,
    flutterEngine: FlutterEngine
) : SensorEventListener {

    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private val rotationVectorSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
    private val magnetometerSensor = sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)

    private var eventSink: EventChannel.EventSink? = null
    private var isListening = false

    // Buffers for sensor data
    private var lastRotationMatrix = FloatArray(9)
    private var lastMagnetometerMagnitude = 0f
    private var lastMagnetometerAccuracy = -1

    init {
        val channel = EventChannel(
            flutterEngine.dartExecutor.binaryMessenger,
            "com.nomatch/orientation_vector"
        )
        channel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
                start()
            }

            override fun onCancel(arguments: Any?) {
                stop()
                eventSink = null
            }
        })
        Log.d("OrientationVectorManager", "[OV-Android] Initialized")
    }

    private fun start() {
        if (isListening) return
        
        if (rotationVectorSensor != null) {
            sensorManager.registerListener(this, rotationVectorSensor, SensorManager.SENSOR_DELAY_FASTEST)
            Log.d("OrientationVectorManager", "[OV-Android] ✅ Listening to rotation vector")
        } else {
            Log.w("OrientationVectorManager", "[OV-Android] ❌ Rotation vector sensor not available")
        }

        if (magnetometerSensor != null) {
            sensorManager.registerListener(this, magnetometerSensor, SensorManager.SENSOR_DELAY_FASTEST)
            Log.d("OrientationVectorManager", "[OV-Android] ✅ Listening to magnetometer")
        }

        isListening = true
    }

    fun stop() {
        if (!isListening) return
        sensorManager.unregisterListener(this)
        isListening = false
        Log.d("OrientationVectorManager", "[OV-Android] ⏹️ Stopped")
    }

    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_ROTATION_VECTOR -> {
                // Convert rotation vector to rotation matrix
                val rotMatrix = FloatArray(9)
                SensorManager.getRotationMatrixFromVector(rotMatrix, event.values)
                lastRotationMatrix = rotMatrix

                // Compute forward direction in world coordinates
                // Device forward = [0, 1, 0] (top of screen in portrait)
                val forwardDevice = floatArrayOf(0f, 1f, 0f)
                val forwardWorld = FloatArray(3)

                // Forward_world = R * forward_device
                for (i in 0..2) {
                    forwardWorld[i] =
                        lastRotationMatrix[i * 3] * forwardDevice[0] +
                        lastRotationMatrix[i * 3 + 1] * forwardDevice[1] +
                        lastRotationMatrix[i * 3 + 2] * forwardDevice[2]
                }

                // Emit event
                val data = mapOf(
                    "fx" to forwardWorld[0].toDouble(),
                    "fy" to forwardWorld[1].toDouble(),
                    "fz" to forwardWorld[2].toDouble(),
                    "magMag" to lastMagnetometerMagnitude.toDouble(),
                    "magAccuracy" to lastMagnetometerAccuracy,
                    "timestamp" to System.currentTimeMillis()
                )
                eventSink?.success(data)
            }

            Sensor.TYPE_MAGNETIC_FIELD -> {
                // Compute magnetometer magnitude and update accuracy
                lastMagnetometerMagnitude = sqrt(
                    event.values[0] * event.values[0] +
                    event.values[1] * event.values[1] +
                    event.values[2] * event.values[2]
                )
                lastMagnetometerAccuracy = event.accuracy
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
        if (sensor.type == Sensor.TYPE_MAGNETIC_FIELD) {
            lastMagnetometerAccuracy = accuracy
        }
    }
}

File: ios/Runner/OrientationVectorManager.swift
import CoreMotion
import Flutter

class OrientationVectorManager: NSObject {
    private let motionManager = CMMotionManager()
    private let magnetometerManager = CMMotionManager()
    private let updateInterval = 0.05 // 20Hz
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private var lastMagnetometerData: CMCalibratedMagneticField?

    func initializeWithChannel(_ channel: FlutterEventChannel) {
        self.eventChannel = channel
        channel.setStreamHandler(self)
        print("[OV-iOS] ✅ OrientationVectorManager initialized")
    }

    private func startMotionUpdates() {
        guard motionManager.isDeviceMotionAvailable else {
            print("[OV-iOS] ❌ Device motion not available")
            return
        }

        motionManager.deviceMotionUpdateInterval = updateInterval
        motionManager.startDeviceMotionUpdates(
            using: .xMagneticNorthZVertical,
            to: .main
        ) { [weak self] _, _ in
            self?.processMotionData()
        }

        print("[OV-iOS] ✅ Started device motion updates (20Hz)")
    }

    private func stopMotionUpdates() {
        motionManager.stopDeviceMotionUpdates()
        print("[OV-iOS] ⏹️  Stopped device motion updates")
    }

    private func processMotionData() {
        guard let motion = motionManager.deviceMotion else { return }

        // Get attitude quaternion
        let att = motion.attitude
        let quat = att.quaternion

        // Convert quaternion to rotation matrix
        let rotMatrix = quaternionToRotationMatrix(
            w: quat.w,
            x: quat.x,
            y: quat.y,
            z: quat.z
        )

        // Forward device in device coordinates = [0, 1, 0] (top of screen)
        let forwardDevice: [Double] = [0, 1, 0]

        // Transform to world: forward_world = R * forward_device
        var forwardWorld: [Double] = [0, 0, 0]
        for i in 0..<3 {
            for j in 0..<3 {
                forwardWorld[i] += rotMatrix[i * 3 + j] * forwardDevice[j]
            }
        }

        // Get magnetometer magnitude from device motion
        let magField = motion.magneticField
        let magMagnitude = sqrt(
            magField.field.x * magField.field.x +
            magField.field.y * magField.field.y +
            magField.field.z * magField.field.z
        )

        // Convert accuracy enum to Int (-1, 0, 1, 2)
        let magAccuracy: Int
        switch magField.accuracy {
        case .uncalibrated:
            magAccuracy = -1
        case .low:
            magAccuracy = 0
        case .medium:
            magAccuracy = 1
        case .high:
            magAccuracy = 2
        @unknown default:
            magAccuracy = -1
        }

        // Emit event
        let data: [String: Any] = [
            "fx": forwardWorld[0],
            "fy": forwardWorld[1],
            "fz": forwardWorld[2],
            "magMag": magMagnitude,
            "magAccuracy": magAccuracy,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000)
        ]

        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(data)
        }
    }

    private func quaternionToRotationMatrix(
        w: Double,
        x: Double,
        y: Double,
        z: Double
    ) -> [Double] {
        // Normalize quaternion
        let norm = sqrt(w * w + x * x + y * y + z * z)
        let qw = w / norm
        let qx = x / norm
        let qy = y / norm
        let qz = z / norm

        // Compute 3x3 rotation matrix from normalized quaternion (row-major)
        let m00 = 1 - 2 * (qy * qy + qz * qz)
        let m01 = 2 * (qx * qy - qw * qz)
        let m02 = 2 * (qx * qz + qw * qy)

        let m10 = 2 * (qx * qy + qw * qz)
        let m11 = 1 - 2 * (qx * qx + qz * qz)
        let m12 = 2 * (qy * qz - qw * qx)

        let m20 = 2 * (qx * qz - qw * qy)
        let m21 = 2 * (qy * qz + qw * qx)
        let m22 = 1 - 2 * (qx * qx + qy * qy)

        return [
            m00, m01, m02,
            m10, m11, m12,
            m20, m21, m22,
        ]
    }

    deinit {
        stopMotionUpdates()
    }
}

extension OrientationVectorManager: FlutterStreamHandler {
    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        self.eventSink = events
        startMotionUpdates()
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        stopMotionUpdates()
        return nil
    }
}
